plutus-core/generators/Language/PlutusCore/Generators/NEAT/Type.hs

```diff
@@ -46,213 +65,350 @@ data TypeBuiltinG
 
 deriveEnumerable ''TypeBuiltinG
 
--- |Convert generated builtin types to Plutus builtin types.
-toTypeBuiltin :: TypeBuiltinG -> Some (TypeIn DefaultUni)
-toTypeBuiltin TyByteStringG = Some (TypeIn DefaultUniByteString)
-toTypeBuiltin TyIntegerG    = Some (TypeIn DefaultUniInteger)
-toTypeBuiltin TyStringG     = Some (TypeIn DefaultUniString)
-
-
--- * Enumerating types
-
 -- NOTE: Unusually, the application case is annotated with a kind.
 --       The reason is eagerness and efficiency. If we have the kind
 --       information at the application site, we can check the two
 --       subterms in parallel, while evaluating as little as possible.
 
-data TypeG n
-  = TyVarG n
-  | TyFunG (TypeG n) (TypeG n)
-  | TyIFixG (TypeG n) (Kind ()) (TypeG n)
-  | TyForallG (Kind ()) (TypeG (S n))
+data TypeG tyname
+  = TyVarG tyname
+  | TyFunG (TypeG tyname) (TypeG tyname)
+  | TyIFixG (TypeG tyname) (Kind ()) (TypeG tyname)
+  | TyForallG (Kind ()) (TypeG (S tyname))
   | TyBuiltinG TypeBuiltinG
-  | TyLamG (TypeG (S n))
-  | TyAppG (TypeG n) (TypeG n) (Kind ())
-  deriving (Typeable, Eq, Show)
+  | TyLamG (TypeG (S tyname))
+  | TyAppG (TypeG tyname) (TypeG tyname) (Kind ())
+  deriving (Typeable, Eq, Show, Functor)
 
 deriveEnumerable ''Kind
 
 deriveEnumerable ''TypeG
 
+type ClosedTypeG = TypeG Z
 
+instance Enumerable tyname => Enumerable (Normalized (TypeG tyname)) where
```

Comment:
I have added a note above this instance. One could follow the approach of `metatheory` and introduce a first order AST with some other kind of normalizer/hereditary substitution rather than the simple small step reduction and parallel substitution implemented here. This would be safer but we already have that in `metatheory`. This seems like a reasonable lightweight approach to use here to me.

Reply:
We could do simple small step reduction *and* have a first-order AST. I'm asking 

> But could we make it a first-order AST and just derive the instance?

because a first-order AST is easier to read and once we have it, we can derive the instance, so we won't need to do that manually and deal with internal NEAT stuff.

Anyway, not too important.